_posts/2016-04-26-aws-and-consul.md

```diff
@@ -0,0 +1,175 @@
+
+Next I will go on to discuss some of the small features of the AWS suite and how we at Sky Betting and Gaming use them to deploy our services.
+
+Finally let's pull it all together and look at how to leverage the power of AWS combined with Consul's clustering.
+
+## Consul Clustering
+A Consul node can run in two different modes. As an agent, or as a server. An agent is similar to a client in other similar architectures, as it's responsible for consuming data from the server. The difference, however, is that agents are also responsible for health checking themselves, and also for discovering other nodes. A server has all the basic functionality of an agent but are also where data is ultimately stored and replicated within the cluster. Although a cluster consists of both agents and servers, only the server nodes store data on disk (ignoring any caching!).
+
+Similar to Zookeeper, Consul works by electing itself a leader within the cluster, assuming that there is a *quorum* of at least 3 nodes in the cluster. This is required, as if there were less, it would be impossible for the election to reach a unanimous result. If the leader fails somehow and the cluster can no longer communicate with it, then a new leader is elected. For this reason, it is best to have a cluster of at least 5 nodes, then we can survive 2 nodes failing with an intact quorum.
+
+The last thing we need to know about for this post is how new nodes join an existing cluster and how to gracefully leave a cluster. A new node knows about it's peers through configuration. You can either pass a list of IP addresses on the command line or better, write the addresses into config. Why is this better? We can then use systemd or upstart or whatever to start Consul each time. 
+
+A node also needs to leave the cluster gracefully, although a cluster will handle an ungraceful exit, it's always cleaner to exit gracefully if it's planned. We do this by including a line like `consul leave` in the systemd service file. That way when the underlying host is shut down, the cluster integrity is maintained.
+
+## EC2 Autoscaling Groups
+For anybody unfamiliar with EC2 Autoscaling I suggest looking into it, it is the single most compelling reason for moving your compute onto AWS. In a nutshell, it allows you to automatically build or destroy EC2 instances based on some criteria, for example, load or simply time of day. Running instances 24/7 in AWS in not particularly cheap, but if you can only use compute on demand, then it becomes very cost effective. 
```

Comment:
Suggest replacing the comma after "into it" with a semi-colon, and changing to "based on some criteria: for example load or, simply, time of day."